### PR TITLE
Moving to async await

### PR DESCRIPTION
**Pull request by Darksonn**
We have been using this library a lot here at omnio and are working on updating our code base to use the new async features. We have already helped update tokio-serial [here][1] and would love to see tokio-modbus updated as well. We internally built an updated version of tokio-modbus, and you can find the changes in this PR.

This pull requests still needs some polishing, but we would like to hear your opinion on the direction we've been taking the library before continuing the implementation, especially regarding the removal of `tokio-proto`, which has a certain impact on the external api:

1. Several methods now require `&mut self` instead of `&self`.
2. These methods also borrow self inside the returned future.

The `tokio-proto` crate previously supplied a layer that uses [interior mutability][2] to put requests in a queue and later [dispatch][3] the results to the appropriate future. As proto is now gone, this dispatching mechanism is also gone, which means that we instead have to rely on mutable borrows to ensure at compile time that at most a single modbus request exists per connection at any time.

The disadvantage of modifying the api in this way is that it can be harder to use the api: previously you were able to just create a bunch of requests and await them all. Now you must create the next request when the preceding one finishes.

Luckily it also turns out that the introduction of async makes the kind of loop described above much easier to write: The introduction of pin makes futures that borrow from the modbus context usable, and the ability to write a loop in async code avoids [certain issues][4] with using `for_each` to create the requests.

```rust
// example of using a loop to read many requests.
let requests_i_want = vec![...];
let results = Vec::with_capacity(requests_i_want.len());
for request in requests_i_want {
    let result = modbus_context.call(request).await?;
    results.push(result);
}
return results;
```

There are also quite a few traits that return a `Pin<Box<...>>`. It would be nice to use associated types in the traits to return non-heap-allocated futures instead, but we have not looked into implementing that yet.

[1]: https://github.com/berkowski/tokio-serial/pull/24
[2]: https://github.com/tokio-rs/tokio-proto/blob/master/src/util/client_proxy.rs#L24
[3]: https://github.com/tokio-rs/tokio-proto/blob/master/src/streaming/pipeline/client.rs#L133
[4]: https://users.rust-lang.org/t/read-from-stream-write-to-file-tokio-await/34150

### CODE REVIEW

**Comment by flosse** [comment]
@Darksonn Thanks a lot for your PR and your work!
Getting rid of `tokio-proto` was definitely one of our main goals (see #18). And getting async/await support is even better :)
Unfortunately we're quite busy right now but I hope we can give you some better feedback soon.
And feel free to change the public API, the `v0.4.0` will have breaking changes anyway ;-)

**Comment by uklotzde** [comment]
I have rebased your branch on master and did some minor cleanup:
https://github.com/uklotzde/tokio-modbus/tree/async2

All tests still work. Could you please check if everything else works as expected? Hopefully nothing got lost along the way. After further adjustments and some tuning you might then reset and force push this branch. We try to keep the version history linear, whenever possible.

I would suggest to exclude the `tcp_server` module in the first step and decide about the final module structure later. Maybe we should also hide it behind a feature flag, because it might only be needed by very few users of this library.

**Comment by not-fl3** [comment]
> I would suggest to exclude the tcp_server module in the first step and decide about the final module structure later. Maybe we should also hide it behind a feature flag, because it might only be needed by very few users of this library.

Hiding "tcp_server" behind feature flag saved client-only users from "net2" dependencie and from all the ex-"tokio-proto" traits, wich is nice! 

PR updated.

**Comment by lovasoa** [comment]
Is there anything preventing this to be merged ? Anything we can do to help ?

**Comment by not-fl3** [comment]
I would propose to publish this as `0.4.0-alpha.1` and continue testing and collecting feedback.

**Comment by flosse** [comment]
> I would propose to publish this as `0.4.0-alpha.1` and continue testing and collecting feedback.

I'll chat with @uklotzde and probably we'll be able to review your PR at the end of the week :)

File: src/client/mod.rs [hunk 1]
```diff
@@ -15,67 +15,77 @@ use crate::slave::*;
...
 
+use std::pin::Pin;
+
 /// A transport independent asynchronous client trait.
-pub trait Client: SlaveContext {
-    fn call(&self, request: Request) -> Box<dyn Future<Item = Response, Error = Error>>;
+pub trait Client: SlaveContext + Send {
+    //fn call(&self, request: Request) -> Pin<Box<dyn Future<Output = Response>>>;
+    fn call<'a>(
+        &'a mut self,
```
Comment (flosse) [inline src/client/mod.rs]: Is there a way to avoid `mut` ?

File: src/client/sync/rtu.rs [hunk 2]
```diff
@@ -18,10 +17,12 @@ pub fn connect_slave(
     settings: &SerialPortSettings,
     slave: Slave,
 ) -> Result<Context> {
-    let mut core = Core::new()?;
-    let handle = core.handle();
-    let serial = Serial::from_path_with_handle(tty_path, settings, &handle.new_tokio_handle())?;
-    let async_ctx = core.run(async_connect_slave(&handle, serial, slave))?;
-    let sync_ctx = Context { core, async_ctx };
+    let mut rt = tokio::runtime::Runtime::new().unwrap();
```
Comment (flosse) [inline src/client/sync/rtu.rs]: Why not map the error instead of `unwrap`?

File: src/client/sync/tcp.rs [hunk 3]
```diff
@@ -15,9 +13,11 @@ pub fn connect(socket_addr: SocketAddr) -> Result<Context> {
 /// gateway that is forwarding messages to/from the corresponding unit identified
 /// by the slave parameter.
 pub fn connect_slave(socket_addr: SocketAddr, slave: Slave) -> Result<Context> {
-    let mut core = Core::new()?;
-    let handle = core.handle();
-    let async_ctx = core.run(async_connect_slave(&handle, socket_addr, slave))?;
-    let sync_ctx = Context { core, async_ctx };
+    let mut rt = tokio::runtime::Runtime::new().unwrap();
```
Comment (flosse) [inline src/client/sync/tcp.rs]: Why not map the error instead of `unwrap`?

File: src/service/tcp.rs [hunk 4]
```diff
@@ -67,16 +74,22 @@ impl Context {
         }
     }
 
-    pub fn call(&self, req: Request) -> impl Future<Item = Response, Error = Error> {
+    pub async fn call(&mut self, req: Request) -> Result<Response, Error> {
```
Comment (flosse) [inline src/service/tcp.rs]: Is there a way to avoid `mut`?

File: src/server/service.rs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    /// Requests handled by the service.
+    type Request;
+
+    /// Responses given by the service.
+    type Response;
+
+    /// Errors produced by the service.
+    type Error;
+
+    /// Process the request and return the response asynchronously.
```
Comment (uklotzde) [inline src/server/service.rs]: "asynchronously"?

**Comment by Darksonn** [comment]
Regarding the mut markers there were added, please see this snippet from the original post:

>... especially regarding the removal of `tokio-proto`, which has a certain impact on the external api:
>
>1. Several methods now require &mut self instead of &self.
>2. These methods also borrow self inside the returned future.
>
> The `tokio-proto` crate previously supplied a layer that uses interior mutability to put requests in a queue and later dispatch the results to the appropriate future. As proto is now gone, this dispatching mechanism is also gone, which means that we instead have to rely on mutable borrows to ensure at compile time that at most a single modbus request exists per connection at any time.

File: Cargo.toml [hunk 6]
```diff
@@ -11,28 +11,30 @@ repository = "https://github.com/slowtec/tokio-modbus"
...
 [dev-dependencies]
 env_logger = "0.6"
+tokio = { version = "0.2", features = ["tcp", "rt-core", "macros"] }
 
 [features]
 default = ["tcp", "rtu", "sync"]
 rtu = ["tokio-serial"]
 tcp = []
 sync = []
+server = ["net2"]
```
[the remark this answers is not in the document]
Comment (lovasoa) [reply Cargo.toml]: I think this is not possible to express without having two distinct `server-tcp` and `server-rtu` features.

File: src/server/service.rs (hunk 5 above)
Comment (lovasoa) [reply src/server/service.rs]: https://en.wiktionary.org/wiki/asynchronously

Comment (uklotzde) [reply src/server/service.rs]: The invocation of this function is synchronous or did I miss anything?

Comment (lovasoa) [reply src/server/service.rs]: Oh, sorry, I wasn't seeing the function declaration that is below your comment.

File: src/codec/rtu.rs [hunk 7]
```diff
@@ -540,16 +541,18 @@ mod tests {
...
-                0x02, // function code
-                0x03, // byte count
-                0x00, // data
-                0x00, // data
-                0x00, // data
-                0x00, // CRC first byte
-                      // missing crc second byte
-            ]);
+            let mut buf = BytesMut::from(
+                &[
```
Comment (uklotzde) [inline src/codec/rtu.rs]: What was the motivation to replace vec! with an array coerced to a slice? Now BytesMut::from will allocate the Vec internally: https://docs.rs/bytes/0.5.3/src/bytes/bytes_mut.rs.html#1013

Doesn't really matter, I'm just asking out of curiosity.

Comment (not-fl3) [reply src/codec/rtu.rs]: I dont actually have an opinion here - `impl From<Vec<u8>>` was removed and I did the simplest thing to fix it. Maybe `from_vec` would be better here as a more explicit alternative.

**Comment by Darksonn** [comment]
> Oh yes... I'm sorry for bothering you.

Don't worry about it! There's also the question of whether or not having them removed is a good idea. Maybe it would be preferable to introduce our own request dispatching?

File: Cargo.toml (hunk 6 above)
[the remark this answers is not in the document]
Comment (not-fl3) [reply Cargo.toml]: I also think that this is not possible without additional features.

File: src/client/sync/rtu.rs (hunk 2 above)
Comment (not-fl3) [reply src/client/sync/rtu.rs]: fixed

File: src/client/sync/tcp.rs (hunk 3 above)
Comment (not-fl3) [reply src/client/sync/tcp.rs]: fixed

**Comment by not-fl3** [comment]
> Don't worry about it! There's also the question of whether or not having them removed is a good idea. Maybe it would be preferable to introduce our own request dispatching?

My arguments on keeping `mut` and not reimplemnting `tokio-proto`'s requests dispatching:
modbus protocol is not multiplexed - we should wait for the response before making new request. With `mut` type system express it just fine. 
If asynchronous requests with interior mutability and requests batching queues needed - it can be made outside of `tokio-modbus`.

File: Cargo.toml (hunk 6 above)
[the remark this answers is not in the document]
Comment (flosse) [reply Cargo.toml]: ok, so never mind ;-)

File: src/codec/rtu.rs (hunk 7 above)
Comment (uklotzde) [reply src/codec/rtu.rs]: My mistake, I forgot to actually build the tests :see_no_evil:

**Comment by flosse** [comment]
@uklotzde The PR now looks ok for me to be merged :)

File: Cargo.toml [hunk 8]
```diff
@@ -11,28 +11,30 @@ repository = "https://github.com/slowtec/tokio-modbus"
...
 
 # Disable default-features to exclude unused dependency on libudev
-tokio-serial = { version = "3.3", optional = true, default-features = false }
+tokio-serial = { version = "4.3.3", optional = true, default-features = false }
+
+net2 = { version = "0.2", optional = true }
 
 [dev-dependencies]
 env_logger = "0.6"
+tokio = { version = "0.2", features = ["tcp", "rt-core", "macros"] }
```
Comment (uklotzde) [inline Cargo.toml]: Only "tcp" and "macros" seem to be required?

File: src/server/service.rs (hunk 5 above)
Comment (not-fl3) [reply src/server/service.rs]: Fixed! Now it is asynchronous again :)

Comment (not-fl3) [reply src/server/service.rs]: Originally I thought that it should not - all the usecases I can find will use `type Future = Ready<..>`. But its an unnecessary API breaking change, so lets keep it async?

File: Cargo.toml (hunk 8 above)
Comment (not-fl3) [reply Cargo.toml]: Confirm, fixed.

File: Cargo.toml [hunk 9]
```diff
@@ -11,28 +11,30 @@ repository = "https://github.com/slowtec/tokio-modbus"
...
+futures = "0.3"
 log = "0.4"
-smallvec = { version = "0.6", default-features = false }
-tokio-io = "0.1"
-tokio-core = "0.1"
-tokio-codec = "0.1"
-tokio-proto = "0.1"
-tokio-service = "0.1"
+smallvec = { version = "1", default-features = false }
+tokio = { version = "0.2", features = ["tcp", "rt-core"] }
```
[the remark this answers is not in the document]
Comment (not-fl3) [reply Cargo.toml]: Did not find any easy ways to avoid "rt-core" here as well. Added most relevant cargo issues to as a comment in Cargo.toml.

**Comment by flosse** [comment]
Rebased and merged in fbd36a400fc5be6e02dec772490ba8ce8ccb9f2e :)
Thank you all!

**Pull request closed without merging**